src/bootchooser.c

```diff
@@ -1375,6 +1375,206 @@ static gboolean efi_get_state(RaucSlot* slot, gboolean *good, GError **error)
+{
+	g_autoptr(GSubprocess) sub = NULL;
+	GError *ierror = NULL;
+	g_autoptr(GBytes) stdout_buf = NULL;
+	gint ret;
+	gsize size;
+	const char *data;
+	gchar *script_name = r_context()->config->custom_script;
+
+	g_return_val_if_fail(cmd, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (bootname)
+		sub = g_subprocess_new(G_SUBPROCESS_FLAGS_STDOUT_PIPE, &ierror, script_name, cmd, bootname, NULL);
+	else
+		sub = g_subprocess_new(G_SUBPROCESS_FLAGS_STDOUT_PIPE, &ierror, script_name, cmd, NULL);
+
+	if (!sub) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to start : %s", script_name);
+		return FALSE;
+	}
+
+	if (!g_subprocess_communicate(sub, NULL, NULL, &stdout_buf, NULL, &ierror)) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to run : %s", script_name);
+		return FALSE;
+	}
+	data = g_bytes_get_data(stdout_buf, &size);
+	// Drop terminating character
+	*ret_str = g_string_new_len(data, size - 1);
+
+	ret = g_subprocess_get_exit_status(sub);
+	if (ret != 0) {
+		g_set_error(
+				error,
+				G_SPAWN_EXIT_ERROR,
+				ret,
+				"%s failed with wrong exit code", script_name);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/* Wrapper for set commands accessing custom script
+ *
+ * @param cmd What to input as command to the custom backend. Mandatory.
+   Valid values:
+   - set-primary
+   - set-state
+ * @param bootname slot.bootname
+ * @param arg extra arguments if needed
+ * @param error Return location for a GError
+ */
+static gboolean custom_backend_set(const gchar *cmd, const gchar *bootname, const gchar *arg, GError **error)
+{
+	g_autoptr(GSubprocess) sub = NULL;
+	GError *ierror = NULL;
+	g_autoptr(GBytes) stdout_buf = NULL;
+	gchar *script_name = r_context()->config->custom_script;
+
+	g_return_val_if_fail(cmd, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (arg)
+		sub = g_subprocess_new(G_SUBPROCESS_FLAGS_NONE, &ierror, script_name, cmd, bootname, arg, NULL);
+	else
+		sub = g_subprocess_new(G_SUBPROCESS_FLAGS_NONE, &ierror, script_name, cmd, bootname, NULL);
+
+	if (!sub) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to start : %s", script_name);
+		return FALSE;
+	}
+
+	if (!g_subprocess_wait_check(sub, NULL, &ierror)) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to run: %s", script_name);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/* Set slot status values */
+static gboolean custom_set_state(RaucSlot *slot, gboolean good, GError **error)
+{
+	GError *ierror = NULL;
+	const gchar *cmd = "set-state";
+	const gchar *state = good ? "good" : "bad";
+
+	g_return_val_if_fail(slot, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (!custom_backend_set(cmd, slot->bootname, state, &ierror) ) {
+		g_propagate_error(error, ierror);
+		return FALSE;
+	}
+
+	return TRUE;
+}
```

Comment:
Can one drop the local ierror variable, the sanity checking of error, and just make the last five lines `return custom_backend_set(cmd, slot->bootname, state, error)`?

Reply:
yes. Included. Thanks